![python image](https://www.python.org/static/img/python-logo@2x.png)

# <center> Python Tutorial -Session 1a</center>
Nirad Banskota

CBGC

(11/16/2023)

## What is Python?

- Python was created by Guido van Rossum, and  first released in 1991.
  - Since then it has gone through multiple changes.  Python 2.0 was released in 2000!<img src="https://gvanrossum.github.io/images/guido-headshot-2019.jpg" align=right width=100></img>
  - Python 3.0 was released in 2008. There are differences between Python 2 and Python 3 
  - In this class, we will be focusing on Python 3.
- Python is a high-level programming language.
   - This is in contrast to a lower-level programming language such as Assembly, which is much closer to the machine language or binary code.
  
- Python is interpretable
  - What this means is that unlike C++, where you need to compile your program, all you need to do is type your code and let the program take care of the rest
  
- Python supports both procedural and object oriented programming
    - These are different paradigms of programming.  Java, for example, is an object oriented programming language.  C is fully procedural.  
- Python is one of the most popular programming languages today
   - Per the TIOBE index, Python is the most popular programming language today
   - R in contrast is ranked lower
- Python can be used for many different things
  - Web Development using frameworks such as Django and Flask
  - Data analysis using packages such as Numpy and Pandas, as well as plotting packages such as Matplotlib and Seaborn
  - General Machine Learning packages such as sklearn
  - Deep learning using packages such as pyTorch and Tensorflow
- Bioinformatics packages are also available
    - Biopython is quite popular
    - For single cell RNA-Seq programming, Scanpy is the second most popular package after Seurat
    - Multiple packages are also available for Spatial Transcriptomics
    - Many Bioinformatics tools have been written in Python
- Python is easy to program.  However, compared to C++ or C, it is slower
    - Some packages have been written in C or C++ such as Numpy.

## <font color="green">Topics to be covered today</font>
 - ### The Basics
     - [Hello World](#print)
     - [Variables](#variables)
     - [If Statement](#if)     
     - [Loops](#loops)
     - [Function](#function)


The plan is to cover all of these topics today.  If for whatever reason, I am unable to, we will continue from where we left off.
As is customary in programming langue classes, we are going to start with <b> Hello World </b>.

This is a simple print command

### <a id="print"> Hello World </a>

In Python 3, the print() function is used for displaying output to the screen.
Note that this is different from python 2, where the parenthesis was optional.
Without parenthesis, an error message gets generated.

In [1]:
print("Hello World!")

Hello World!


There is also the input() function, which asks for input from the user.  

In [2]:
input("What would you like served for dinner, sir?")

What would you like served for dinner, sir? Rice


'Rice'

You add comments by using using "#" as follows:

In [6]:
# This is a comment


If you highlight a section of code and present ctrl and "/" , then you can do a block comment.

In [2]:
# 1
# 2
# 3
# 4

<b>Exercise 0:</b>

Use the print() function to print "Hello NIA!" to the screen.  Use the cell below

<b>Time: 10 seconds.</b>


------------------------------------------------------------------


### <a id="variables">Assigning to variables </a>

Let's suppose we want to store the word from the input statement above for future use.  How do we do that?
We simply need to assign it to a variable. So basically, a variable is a way for storing values.

The way to do this is with the syntax variable_name=<value or object, etc>  (Note without the "<" or ">")


Rules to follow:
 - Cannot start with number
 - Special Characters do not work
 - The underscore works everywhere
 - Note that Hyphens do not work


Here are some examples of assigning values to variables

In [5]:
a=3
b="apple"
c="Hello World"
_d="Que Pasa"
j3="This is j3"
print(a)
print (b)
print(c)
print(_d)
print(j3)

3
apple
Hello World
Que Pasa
This is j3


<b>The following examples will not work: </b>

<code>
-j=3
@$*$k="apple"
3j="cat"
</code>

Now let's make a more interactive script using print and input functions

In [7]:
j=input("Hello.  What would you like for dinner? ")
print("You want",j)

Hello.  What would you like for dinner?  Rice


You want Rice


In the script above, we first use the input() function to ask the user a question.  
Next, we store it to a variable that we call j.  
Finally, we use the print statement to print that variable.
Here is an alternative way to do the same thing as above.  

##### Exercise 1:<br>
<i>
    What I want you to do is input <b> 2 numbers </b> from the user.  

Assign the first number to the variable <b>apples</b>

Assign the second number to the variable <b>oranges</b>.

Print the number of apples and oranges to  the screen using the print statement.</i>


An example output:

Enter the number of apples: 10

Enter the number of oranges: 15

You have 10 apples and 10 oranges

Note: The sentence you want to print and the variable can be separated by commas as follows:

a=3

b=10

print("The value of a is",a," and the value of b is ",b)

Time: <b> 2 minutes </b>

----------------------

There is another way to embed a variable into the print function, as I demonstrate below:

In [8]:
j=input("Hello.  What would you like for dinner? ")
print(f"You want {j}")

Hello.  What would you like for dinner?  Meat


You want Meat


The above method is called f-strings, and it is used for formatting strings.  Briefly, it replaces variables within the {} brackets by the actual value. Some more examples below:

In [9]:
print(f"{3*4}")
a="apple"
b="ball"
print(f"A for {a}. B for {b}")

12
A for apple. B for ball


##### Exercise 2 <br>
<i> Repeat Exercise 1.  However this time use f-strings.</i>

As a reminder, here is what exercise 1 is about:

<i>
    What I want you to do is input <b> 2 numbers </b> from the user.  

Assign the first number to the variable <b>apples</b>

Assign the second number to the variable <b>oranges</b>.

Print the number of apples and oranges to  the screen using the print statement.</i>



Time: <b> 2 minutes </b>

-----------------------------------

Now let's describe some simple mathematical operations.  

In [23]:
print(3+3)
print(4+4)
print(5-4)
print(6*4)
print(7/4)
print(9/4)
print(7//4)
print(9//4)
print(2**3)
print (9%4)
print(10%5)
print(12%5)

6
8
1
24
1.75
2.25
1
2
8
1
0
2


The "+" is used for adding, "-" for subtracting "*" for multiplication, "/" for division.  The double "//" is used for floor division. In other words, while 7/4 is 1.75, the double // will ignore the .75 (values after decimal) and will print just the 1 in front.The "%" operator is the modulus.  It returns the remainder.  In other words, when you divide 9 by 4, it will return 1, because 4 times 2 is 8, and 9 minus 8 is 1. 10%5 does not have a remainder, so it returns 0. Some more examples of the modulus operator below:

In [24]:
print(5%10)
print(8%3)
print(12%3)

5
2
0


Now let's introduce some other assignment operators.

In [30]:
a=3
print(f'{a}')
a+=3
print(f'{a}')
a-=3
print(f'{a}')
a=a+3
print(f'{a}')
a=a-3
print(f'{a}')
a*=3
#Above is equivalent to a=a*3
print(f'{a}')
a/=3
#Above is equivalent to a=a/2
print(f'{a}')


3
6
3
6
3
9
3.0


Flipping values of variables.  A neat little trick.

In [10]:
a=3;b=8
print(f"a is {a}, while b is {b}")
a,b=b,a
print(f"a now is {a}, while b now is {b}")

a is 3, while b is 8
a now is 8, while b now is 3


In [31]:
a=3;b=8;c=9
a,b,c=b,c,a
print(f"{a},{b},{c}")

8,9,3


Now let's introduce the type command.

In [11]:
print(type(a))
jackal="jackal"
print(type(jackal))
flt=3.0
print(type(flt))
b=True
print(type(b))
print(type(1))

<class 'str'>
<class 'str'>
<class 'float'>
<class 'bool'>
<class 'int'>


The type command can be used to find out the type of the variable object.  Since jackal stores a string, the class of that variable is "str".  Similarly, flt is a decimal number, and hence, it is a float.  You are also able to convert a string to an int, or an int to a float, etc.

In [33]:
a=3
k=float(a)
print(k)
print(type(k))
l=str(a)
print(type(l))

3.0
<class 'float'>
<class 'str'>







<b> Exercise 3 </b>

<i>  Input 2 numbers from the user.  Assign the first number to variable a.</i>
    
<i> Assign the second number to variable b. </i>
    
<i>    Divide a by b. </i>
    
<i>    Print the final number. </i>



Example:

Enter Number 1 :10

Enter Number 2: 5

Number 1 divided by Number 2 is 2.0



Time: <b> 2 minutes </b>

------------------------------------------

You can concatenate 2 strings as follows:

In [12]:
a="apple"
b="ball"
print(a+b)

appleball


The above commands show that if the variables are strings, then the plus operator can be used for concatenating strings.  The multiplication operator can be used for repeating the same string multiple times, as can be seen below:

In [13]:
a="apple"
b=a*3
print(b)

appleappleapple


In [14]:
a="3"
print(a*3)
print(int(a)*3)

333
9


What the above commands demonstrate is that if the number is stored as a string, then the multiplication operator will just repeat the string that many times.  So what you need to do is convert the string to integer(or float) before calling your script. 

### <a id="if"> The if-else Statement </a>

The if-else statement is used to test whether a statement is true or false.  For example:

- if a certain statement is true, print it out.
- Otherwise(i.e. else) print something else.

The else is optional.

It should be noted that you may also want to compare a value multiple different ways.  For example:
- If a is equal to b and a is greater than c then do this.
- Otherwise(else) do this.

Thus you need logical operators such as <b>and</b>, <b>or</b> and <b>not</b>.

You may also have multiple if statements, which you can chain as follows:
- If a is equal to b then do this.
- else if a is equal to c then do this.
- else if a is equal to j then do this.
- else (if none of the statements above are true) do this


First, let's talk about the <b>comparison</b> operators.  The comparison operators that I will introduce are:

    1)== Equal.  Test for equality.
    2)!= Not equal.  Test for inequality.
    3)< Less Than.  Is a less than b?
    4) > Greater Than.  Is a greater than b?
    5) <= Less than or equal to.  Is a less than or equal to b?
    6) >= Greater than or equal to. Is a greater than or equal to b?
    
Here are some examples:

In [15]:
a=3
if a==3:
    print(a)
# assign 6 and do a not equal
a=6
if a!=3:
    print(a)
if a>5:
    print(a)
if a>=6:
    print(a)
if a<=6:
    print(a)
if a<10:
    print(a)



3
6
6
6
6
6


##### Exercise 4:
<i>  Enter a number (integer) from the user.</i>
    
<i>Print whether the number is odd or even.
    
Example:
    
    Enter a Number:  3
    The number is Odd
 
Note:  Even numbers are 2, 4, 6, 8....16,18...to infinity (i.e. divisible by 2)  Odd numbers are 1,3,5,7,9, to infinity.    
Time:  <b> 3 minutes </b>
    
    

------------------------------

Next,  we have the logical operators


if a=TRUE <b>and</b> b=TRUE --> True   (what I mean is that if the 2 conditions are True, only then will the code run)

if a=TRUE and b=FALSE --> FALSE

if a=FALSE and b=TRUE --> FALSE

if a=TRUE <b>or</b> b=TRUE --> TRUE

if a=TRUE or b=FALSE--> TRUE

if a=FALSE or b=TRUE--> TRUE

if a=FALSE or b=FALSE--> FALSE

if <b>not</b> a:  True


In [16]:
a=3;b=6
#Comparison 1
if a==3 and b==6:
    print("Comparison 1: a is 3 AND b is 6")
#Comparison 2:
if a==3 and b!=6:
    print("Comparison 2: a is 3 AND b is not 6")
#Comparison 3
if a==3 and b!=7:
    print("Comparison 3: a is 3 and b is not 7")
#Comparison 4:
if a==3 or b==7:
    print("Comparison 4: Either a is 3 or b is 7")
# Comparison 5:
if a==8 or b==9:
    print("Comparison 5: Either a is 8 or b is 9")

Comparison 1: a is 3 AND b is 6
Comparison 3: a is 3 and b is not 7
Comparison 4: Either a is 3 or b is 7


The and comparison means that both statements need to be correct.
The or statement means that one of two statements need to be correct.
The comparison operators can be combined as follows


In [17]:
a=3;b=3;c=9
if a==3 and (b==3 or c==8):
    print("a is 3 and either b is 3 or c is 8")

a is 3 and either b is 3 or c is 8


Finally, there is also the not operator

In [18]:
a=3
if not a==3:
    print ("A is not 3")
if not a==5:
    print("A is not 5")

A is not 5


Now let's talk about the if else statement.  The else is optional, but it executes only if the "if statement" is false.

In [20]:
a=3
if a>3:
    print("a is greater than 3")
else:
    print("A is not greater than 3")

A is not greater than 3


Here is how the script is processed.  We first assign 3 to the variable a.  The if statement first looks to see if a is greater than 3.  This will be processed as FALSE.  If the statement was "if a>=3" then it would have been True. Then it goes to the else statement and executes the print statement.
The elif statement can be used for chaining multiple <i>if</i> statements, as shown below.

In [21]:
a=3
if a>3:
    print("A is greater than 3")
elif a<3:
    print("A is less than 3")
elif a>10:
    print("A is greater than 10")
elif a<-10:
    print("A is less than -10")
else:
    print("A equals 3")

A equals 3


Here the first statement is False, because a is not greater than 3. The elif statement will also be process as false, as a is not less than 3.  Finally, the last else statement is executed.  Thus the script processes to 3.

##### Exercise 5
<i> Let the user enter 2 numbers.<i>
    
<i> If both numbers are even, print  "Both numbers are even!" </i>
    
<i> If both numbers are odd, print "Both numbers are odd!" </i>
    
<i> If one number is odd and one number is even, print "One number is even and one number is odd!" </i>
    
Example:
    
        Enter Number 1: 10
    
        Enter Number 2:  15
    
        One number is even and one number is odd!
    
Time:  <b> 3 minutes </b>

--------------------------------------------------------------------------------

### <a id="loops"> Loops in Python </a>

Let's say you want to repeat a task 10 times.  
One way to do this is to write 10 statements.  However, this is tedious.
If the task is repititive, then you want a way to iterate through the tasks.
This is what a loop is used for.  

First let's discuss the <b>for loop</b>, which is the most common type of loop.  

Here is how to use a for loop.

In [48]:
for i in range(10):
    print(i)

0
1
2
3
4
5
6
7
8
9


If you want to print 10 numbers, the range(10) will print 10 numbers. However, it should be noted that unlike R, python is 0-indexed.  So while the for loop prints 10 numbers, it starts from 0 and ends at 9.  

In [26]:
for i in range(1,11):
    print(i)

1
2
3
4
5
6
7
8
9
10


In [27]:
for i in range(0,10,2):
    print(i)

0
2
4
6
8


In the above line, the script starts at 0, and ends at 10(subtract 1 from 1).  And the 2 specifies that we want to skip by 2.  What if I want to print from 10 to 1 in reverse direction?

In [28]:
for i in range(10,0,-1):
    print(i)

10
9
8
7
6
5
4
3
2
1


In [29]:
for i in range(10,0,-2):
    print(i)

10
8
6
4
2


There is an option Else keyword that can be used to specify that the for loop has ended

In [30]:
for i in range(10):
    print(i)
else:
    print("Done")

0
1
2
3
4
5
6
7
8
9
Done


The for loop can be used with the break keyword to get out of the loop.
The continue keyword can be used to skip a loop if it meets a condition

In [31]:
for i in range(10):
    if i==5:
        break
    print(i)


0
1
2
3
4


In [32]:
for i in range(10):
    if i==5:
        continue
    print(i)
    

0
1
2
3
4
6
7
8
9


As can be seen above, the value 5 is skipped due to the continue keyword

In [33]:
for i in range(10):
    if i%2==0:
        continue
    print(i)

1
3
5
7
9


In the code above, if the remainder is 0, we are going to skip that iteration.  This results in only odd numbers being printed.
Let us try to print only the prime numbers.  
As a reminder, a prime number is a number that is only divisible by itself and 1.  Some examples of prime numbers are 2,3,5,7,11, etc. Let's find all the prime numbers from 1 to 30.


In [34]:
for i in range(1,31):
    prime=True
    if i==1:
        continue
    if i==2:
        print(f"{i} is prime!")
        continue
    else:
        for j in range(2,i):
            if i%j==0:
                prime=False
    if prime==True:
        print(f"{i} is prime!")
            

2 is prime!
3 is prime!
5 is prime!
7 is prime!
11 is prime!
13 is prime!
17 is prime!
19 is prime!
23 is prime!
29 is prime!


This is a brute force method for calculating prime numbers.  There are better algorithms for finding prime numbers.
Also, a nested loop is a loop where there is a loop within a loop.
Theoretically, to calculate a  the second loop does not need need to go higher than the square root of the value i. 

Another example of a nested loop to clarify how it works:

In [58]:
for i in range(5):
    for j in range(10,15):
        print(i,j)

0 10
0 11
0 12
0 13
0 14
1 10
1 11
1 12
1 13
1 14
2 10
2 11
2 12
2 13
2 14
3 10
3 11
3 12
3 13
3 14
4 10
4 11
4 12
4 13
4 14


The way it works is that the internal loop(j) gets executed until we get to 14 at i=0. Then the value of i goes up by 1. The internal loop(j) needs to run again until the value 14. And so on and so forth. 

##### Exercise 6

<i> Print 2 numbers from 1 to 5 that will have the following pattern:</i>
    
    1  5
    1  4
    1  3
    1  2
    2  5
    .
    .
    .
    5 1
   <i> Use for loop </i>
   
<i>If two values are the same, it is skipped. </i>

Notice that 1  1 is not printed.
Time:<b> 2 minutes </b>

---------------------

The for loop can also be used to print individual characters in a string:

In [35]:
for i in "abcd":
    print(i)

a
b
c
d


##### Exercise 7:

<i>An amino acid is encoded by 3 nucleotides called a codon. Some codons will code for the same amino acid.  However, we will ignore this.  
Print all possible codons.</i>
Here are the 4 nucleotides: 

1) A

2) T(or U)

3) G

4) C

Time: <b> 3 minutes </b>


-----------------------------------------------------------------------

Now let's introduce the <b>while loop</b>.  
The while loop will run endlessly as long as the conditional statement within it is true.
For example:
while the value of a is less than 3, keep doing this.

<div class="alert alert-block alert-danger">"Note that the while loop can run endlessly unless you specify how to  break out of the loop.</div>

In [36]:
a=1
while(a<11):
    print(a)
    a+=1
    

1
2
3
4
5
6
7
8
9
10


The way the while loop works is that we need to first specify the value of a as 1. While this value is less than 11(meaning up to 10), the while loop can keep running.  Within the loop, we increment the value of a by 1. We can also easily step by 2 if we increment by 2.

In [37]:
a=1
while(a<11):
    print(a)
    a+=2


1
3
5
7
9


In [38]:
a=1
while(not(a>10)):
    print(a)
    a+=1

1
2
3
4
5
6
7
8
9
10


In [39]:
a=5
while(a<10):
    b=15
    while(b<20):
        print(a,b)
        b=b+1
    a=a+1

5 15
5 16
5 17
5 18
5 19
6 15
6 16
6 17
6 18
6 19
7 15
7 16
7 17
7 18
7 19
8 15
8 16
8 17
8 18
8 19
9 15
9 16
9 17
9 18
9 19


##### Exercise 8
<i> In this exercise, you will create a game called guess the number. The number that is guessed will be a random number from 0 to 100. This code is provided. </i>
<i> Your task is to fill the remaining code.  If the number you guess is smaller than the true number, print "too small".  If it is bigger, print too big </i>
<i> If the number is guessed correctly, print "Correct".</i>
<i> Give the user the ability to quit the game any time they want (use the q keyword to quit).</i>

Example:

Guess what number I am thinking of(q to quit):  90

Too big!

Guess what number I am thinking of(q to quit) : 10

Too small!

Guess what number I am thinking of(q to quit) :  50

Correct!

Time: <b> 5 minutes </b>


In [40]:
from random import randint
true_number=randint(0,100)
###Add your loop here


----------------------------------------------------------------

### <a id="function">Functions </a>

A function is a way of storing a block of code so that you can reuse it.  
  - In python, the def function is used for declaring a function.
  - The def keyword is then followed by the name of the function.  In the example below, the function name is F_C_converter.
After the function name are the variables that the function will accept.  
-  The example below has two variables: Temp for temperature, and the  Unit. Is the temperature value in Fahrenheit or Celsius?  
-  We can also declare a default value.  In the example below, the default value is "Fahrenheit".  What that means is that if you do not provide a unit, the script below will assume that the unit of Temperature is Fahrenheit.  The program will then proceed to convert it to Celsius.
- The return keyword is used to return the value (or values).  This value can be stored in a variable if needed.

In [41]:
def F_C_converter(Temp,unit="Fahrenheit"):
    if unit=="Fahrenheit":
        Celsius=((Temp)-32)*(5/9)
        return Celsius
    elif unit=="Celsius":
        Fahrenheit=((Temp*9/5)+32)
        return Fahrenheit
    else:
        return "Error: The unit can only be Celsius or Fahrenheit"

In [42]:
print(F_C_converter(100))
print(F_C_converter(32))

37.77777777777778
0.0


In [43]:
print(F_C_converter(37.77778,"Celsius"))
print(F_C_converter(0,"Celsius"))

100.000004
32.0


In [44]:
print(F_C_converter(0,"Kelvin"))

Error: The unit can only be Celsius or Fahrenheit


If you want more clarity, you can also annotate your variables in a function. In the example below, you expect the input variable to be an integer, and the output to be string

In [45]:
def example_function(a:"int")->"str":
    b=str(a)
    print(f"The type is {type(b)}")

In [46]:
example_function(3)

The type is <class 'str'>


##### Exercise 9
<i> Create a function that takes as input  DNA or CDNA sequence and converts it to to RNA sequence.  </i>

<i>A DNA has nucleotides A,T,G,C/CDNA has A,T,G,C, whereas RNA has A,U,G,C. </i>

Time: <b> 3 minutes </b>

In [71]:
def convert_to_RNA(DNA):
   ##code goes here
    return seq